torch/_inductor/memory.py

```diff
@@ -0,0 +1,525 @@
+            if buf.mpi.outdegree == 0:
+                node.mpi.memory_to_free += buf.mpi.size_free
+
+    # schedule nodes one at a time
+    schedule: List[BaseSchedulerNode] = []
+    while nodes_to_schedule:
+        # select a node to schedule:
+        selected_node = min(
+            nodes_to_schedule,
+            key=lambda node: (
+                max(live_memory + node.mpi.size, max_memory),
+                node.mpi.size - node.mpi.memory_to_free,
+                node.mpi.index,
+            ),
+        )
+        nodes_to_schedule.remove(selected_node)
+        schedule.append(selected_node)
+
+        # update memory usage
+        live_memory += selected_node.mpi.size
+        max_memory = max(max_memory, live_memory)
+        live_memory -= selected_node.mpi.memory_to_free
+
+        # update successor nodes and nodes_to_schedule
+        for succ_node in selected_node.mpi.succ_nodes:
+            assert succ_node.mpi.indegree > 0
+            succ_node.mpi.indegree -= 1
+            if succ_node.mpi.indegree == 0:
+                nodes_to_schedule.add(succ_node)
+
+        # update predecessor nodes
+        for buf in selected_node.mpi.pred_buffers:
+            assert buf.mpi.outdegree > 0
+            buf.mpi.outdegree -= 1
+            if buf.mpi.outdegree == 1:
+                for succ_node in buf.mpi.succ_nodes:
+                    succ_node.mpi.memory_to_free += buf.mpi.size_free
+
+    return schedule
+
+
+def topological_sort_bfs(nodes: List[BaseSchedulerNode]) -> List[BaseSchedulerNode]:
+    """
+    A BFS topological sort that selects nodes whose depednecies are executed
+    the earliest. This follows a FIFO idea.
+    """
+
+    @dataclasses.dataclass
+    class HeapElement:
+        priority: List[int]
+        node: BaseSchedulerNode
+
+        def __lt__(self, other: HeapElement) -> bool:
+            if self.priority == other.priority:
+                return self.node.mpi.index < other.node.mpi.index
+            return self.priority > other.priority
+            return self.priority < other.priority
+
+    def _node_priority(node: BaseSchedulerNode) -> List[int]:
+        assert node.mpi.indegree == 0
+        ids = sorted({pred_node.mpi.index for pred_node in node.mpi.pred_nodes})
+        ids.append(node.mpi.index)
+        return ids
+
+    # compute nodes' number of unmet dependencies (for schedulability)
+    # initialize the list of nodes ready to be scheduled
+    nodes_to_schedule: List[HeapElement] = []
+    for t, node in enumerate(nodes):
+        node.mpi.index = t
+        # note that .unmet_dependencies could have deps with the same name
+        # and in that case, it should only be counted once
+        node.mpi.indegree = len(node.mpi.pred_nodes)
+        if node.mpi.indegree == 0:
+            heapq.heappush(nodes_to_schedule, HeapElement(_node_priority(node), node))
+
+    # schedule nodes one at a time
+    schedule: List[BaseSchedulerNode] = []
+    while nodes_to_schedule:
```

Comment:
same here :)

Reply:
I think I will also add a `verify_order` function to make sure the returned order is correct.